### PR TITLE
Register numel, dim, get_device, storage_offset, is_contiguous in native_functions.yaml

### PR DESCRIPTION
**Pull request by aorenste**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #177200

These fast-path functions were previously excluded from native_functions.yaml
to avoid dispatcher overhead, which required duplicate manual JIT operator
registrations in register_prim_ops.cpp. With manual_cpp_binding: True, the
functions get schema/dispatcher/JIT registration via codegen while the
fast-path C++ methods on TensorBase remain unchanged.

Follows the pattern established by is_complex, is_floating_point, etc.
The prim:: functions (is_cuda, is_sparse, is_mkldnn, is_quantized) are
left for a follow-up since they require JIT frontend and interpreter changes.

Partially addresses https://github.com/pytorch/pytorch/issues/30774

Authored with Claude.

### CODE REVIEW

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -118,14 +118,14 @@ struct Var {
   }
 };
 
-static inline bool is_contiguous(const Tensor& t) {
+static bool is_contiguous_in_any_format(const Tensor& t) {
```
Comment (aorenste) [inline aten/src/ATen/native/Normalization.cpp:121]: This rename was needed because it collided with the name that's now visible in a header file.

File: .lintrunner.toml [hunk 2]
```diff
@@ -187,6 +187,11 @@ exclude_patterns = [
     'aten/src/ATen/cpu/FlushDenormal.cpp',
     'aten/src/ATen/cpu/vml.h',
     'aten/src/ATen/CPUFixedAllocator.h',
+    # interned_strings.h includes the generated aten_interned_strings.h which
+    # #errors under TORCH_ASSERT_ONLY_METHOD_OPERATORS. The .cpp file #undefs
+    # the macro before including the header, but clang-tidy lints the .h
+    # standalone using the .cpp's compile flags, so the #error always fires.
+    'aten/src/ATen/core/interned_strings.h',
```
Comment (aorenste) [inline .lintrunner.toml:198]: This addresses a pre-existing lint that was firing on interned_strings.h simply because I touched the file.

**Review by Skylion007 — APPROVED** [review]

**Review by eellison — COMMENTED** [review]
If the motivation is just for jit, I'm not sure I would land this pr. What are the other benefits ?

**Comment by eellison** [comment]
I'll defer to @ezyang or @albanD on that. agreed some benefit to standardization. but also churn..

**Comment by aorenste** [comment]
@albanD Benchmark results:
Benchmark: P2263659973

  JIT path (10k calls per measurement, the path that changed)

```
┌────────────────┬───────────┬─────────┬───────┐
│       Op       │ Base (us) │ PR (us) │ Delta │
├────────────────┼───────────┼─────────┼───────┤
│ numel          │ 424.2     │ 392.5   │ -7.5% │
├────────────────┼───────────┼─────────┼───────┤
│ dim            │ 356.3     │ 344.9   │ -3.2% │
├────────────────┼───────────┼─────────┼───────┤
│ get_device     │ 414.2     │ 395.6   │ -4.5% │
├────────────────┼───────────┼─────────┼───────┤
│ storage_offset │ 393.8     │ 393.2   │ ~0%   │
├────────────────┼───────────┼─────────┼───────┤
│ is_contiguous  │ 839.7     │ 814.8   │ -3.0% │
└────────────────┴───────────┴─────────┴───────┘
```
Eager path (sanity check, should be unchanged)
```
┌────────────────┬───────────┬─────────┬─────────────┐
│       Op       │ Base (ns) │ PR (ns) │    Delta    │
├────────────────┼───────────┼─────────┼─────────────┤
│ numel          │ 107.0     │ 108.1   │ ~0%         │
├────────────────┼───────────┼─────────┼─────────────┤
│ dim            │ 84.0      │ 77.2    │ -8% (noise) │
├────────────────┼───────────┼─────────┼─────────────┤
│ get_device     │ 79.3      │ 78.4    │ ~0%         │
├────────────────┼───────────┼─────────┼─────────────┤
│ storage_offset │ 87.3      │ 82.8    │ -5% (noise) │
├────────────────┼───────────┼─────────┼─────────────┤
│ is_contiguous  │ 100.0     │ 98.0    │ ~0%         │
└────────────────┴───────────┴─────────┴─────────────┘
```
  batch_norm
```
┌─────────────┬───────────┬─────────┬───────┐
│             │ Base (ms) │ PR (ms) │ Delta │
├─────────────┼───────────┼─────────┼───────┤
│ BatchNorm2d │ 28.49     │ 27.46   │ -3.6% │
└─────────────┴───────────┴─────────┴───────┘
```
Takeaway

No regressions anywhere. The JIT path is slightly faster across the board — likely because the codegen-registered operators have a more direct dispatch path than the old hand-written lambdas in register_prim_ops.cpp (less stack manipulation with pop/push/peek/drop/pack). Eager path is unchanged as expected. Batch norm shows a small improvement, though that's likely within run-to-run variance.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -3295,6 +3295,42 @@
   device_guard: False
   manual_cpp_binding: True
 
+- func: numel(Tensor self) -> int
+  variants: method
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:3299]: Does the variants: flag has any impact here since we're doing manual binding?

**Review by albanD — APPROVED** [review]
Small nit but sounds ok!

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (aorenste) [reply aten/src/ATen/native/native_functions.yaml:3299]: A couple things:
1. This matches the pattern used by existing manual binding functions (is_complex, is_floating_point, etc).
2. manual_cpp_binding renames the generated bindings (from 'foo' to '__dispatch_foo') but they're still generated. Nobody uses them so it doesn't really matter - but they are generated (in both method or function variants) - so we'd be generating incorrect (unused) bindings without this.
